File: Models/Contract.py
from datetime import datetime


class Contract:
    """
    Třída reprezentující smlouvu. Třída obsahuje všechny atributy, které obsahuje smlouva v registru smluv.
    Atributy nejsou povinné a proto mají nastavenou výchozí hodnotu na None.
    """

    def __init__(self, contract_id: int = None, external_id: str = None, version_id: str = None, link: str = None,
                 date_published: datetime = None, ministry_name: str = None, ministry_data_box: str = None,
                 ministry_ico: str = None, ministry_address: str = None, ministry_department: str = None,
                 ministry_payer_flag: str = None, supplier_name: str = None, supplier_date_box: str = None,
                 supplier_ico: str = None, supplier_address: str = None, supplier_department: str = None,
                 supplier_receiver_flag: str = None, purpose: str = None, date_agreed: datetime = None,
                 contract_number: str = None, approved: str = None, amount_without_dph: float = None,
                 amount_with_dph: float = None, amount_different_currency: float = None, currency: str = None,
                 hash_value: str = None, link_pdf: str = None, attachments: list = None, valid: str = None,
                 linked_record:str = None):
        """
        Konstruktor smlouvy. Je navrhnut tak, aby do něj bylo možné uložit co nejvice dat.
        :param contract_id: Identifikátor smlouvy v databazi.
        :param external_id: Identifikátor smlouvy v registru smluv.
        :param version_id: Verze záznamu.
        :param link: url Odkazující na záznam v registu smluv.
        :param date_published: Datum zvěřejnění.
        :param ministry_name: Celé jméno ministerstva.
        :param ministry_data_box: Datová schránka ministerstva.
        :param ministry_ico: IČO ministerstva.
        :param ministry_address: Adresa ministerstva.
        :param ministry_department: Odbor ministerstva.
        :param ministry_payer_flag: Flag, který označuje, že je ministerstvo plátce ve vztahu ke smlouvě.
        :param supplier_name: Celé jméno dodavatele.
        :param supplier_date_box: Datová schránka dodavatele.
        :param supplier_ico: IČO dodavatele.
        :param supplier_address: Adresa dodavatele.
        :param supplier_department: Odbor nebo část dodavatele.
        :param supplier_receiver_flag: Flag, který označuje, že je dodavatel příjmce ve vztahu ke smlouvě.
        :param purpose: Předmět smlouvy.
        :param date_agreed: Datum uzavření.
        :param contract_number: Číslo smlouvy (nemusí se nutně jednat o číslo).
        :param approved: Jméno osoby, které smlouvu schválila.
        :param amount_without_dph: Částka bez DPH.
        :param amount_with_dph: Částka s DPH.
        :param amount_different_currency: Částka v cizí měně.
        :param currency: Cizí měna.
        :param hash_value: Hash hodnota smlouvy.
        :param link_pdf: URL odakzující na smlouvu v pdf.
        :param attachments: list příloh. Pokud žádné nejsou, může být None.
        :param valid: Označení, zda je záznam platný.
        :param linked_record: Identifikátor smlouvy, ke které je tato smlouva navázána.
        """
        self.contract_id = contract_id
        self.external_id = external_id
        self.version_id = version_id
        self.link = link
        self.date_published = date_published
        self.ministry_name = ministry_name
        self.ministry_data_box = ministry_data_box
        self.ministry_ico = ministry_ico
        self.ministry_address = ministry_address
        self.ministry_department = ministry_department
        self.ministry_payer_flag = ministry_payer_flag
        self.supplier_name = supplier_name
        self.supplier_date_box = supplier_date_box
        self.supplier_ico = supplier_ico
        self.supplier_address = supplier_address
        self.supplier_department = supplier_department
        self.supplier_receiver_flag = supplier_receiver_flag
        self.purpose = purpose
        self.date_agreed = date_agreed
        self.contract_number = contract_number
        self.approved = approved
        self.amount_without_dph = amount_without_dph
        self.amount_with_dph = amount_with_dph
        self.amount_different_currency = amount_different_currency
        self.currency = currency
        self.hash_value = hash_value
        self.link_pdf = link_pdf
        self.attachments = attachments
        self.valid = valid
        self.linked_record = linked_record

    def __str__(self):
        return f"Contract[{self.contract_id}]: m_name: {self.ministry_name}, m_ico: {self.ministry_ico}, " \
               f"s_name: {self.supplier_name}, s_ico: {self.supplier_ico}, agreed: {self.date_agreed}, " \
               f"amount: {self.amount_with_dph}"

File: frontend/src/views/Contract/Contract.js
import React, {useEffect, useState} from "react";
import {useParams} from "react-router-dom";
// @material-ui/core components
import { makeStyles } from "@material-ui/core/styles";
// core components
import GridItem from "components/Grid/GridItem.js";
import GridContainer from "components/Grid/GridContainer.js";
import Card from "components/Card/Card.js";
import CardHeader from "components/Card/CardHeader.js";
import CardBody from "components/Card/CardBody.js";
import Link from "@material-ui/core/Link";
import InvoiceDialog from "components/InvoiceDialog/InvoiceDialog";
import {matcher_api_url} from "variables/general"
import {Button, IconButton} from "@material-ui/core";
import LinearProgress from "@material-ui/core/LinearProgress";
import Table from "components/Table/Table.js";
import {FindInPage} from "@material-ui/icons";
import Icon from "@material-ui/core/Icon";
import CardIcon from "components/Card/CardIcon";
import InfoColumn from "components/InfoColumn/InfoColumn";
import Tooltip from "@material-ui/core/Tooltip";
import Typography from "@material-ui/core/Typography";

const styles = {
  cardCategoryWhite: {
    color: "rgba(255,255,255,.62)",
    margin: "0",
    fontSize: "14px",
    marginTop: "0",
    marginBottom: "0"
  },
  cardTitleWhite: {
    color: "#FFFFFF",
    marginTop: "0px",
    minHeight: "auto",
    fontWeight: "300",
    fontFamily: "'Roboto', 'Helvetica', 'Arial', sans-serif",
    marginBottom: "3px",
    textDecoration: "none"
  },
  cardCategoryBlack: {
    color: "rgba(0,0,0,.62)",
    margin: "0",
    fontSize: "14px",
    marginTop: "0",
    marginBottom: "0"
  },
  cardTitleBlack: {
    color: "#000000",
    marginTop: "0px",
    minHeight: "auto",
    fontWeight: "300",
    fontFamily: "'Roboto', 'Helvetica', 'Arial', sans-serif",
    marginBottom: "3px",
    textDecoration: "none"
  },
  whiteIcon:{
    color: "#FFFFFF"}
};

const useStyles = makeStyles(styles);

export default function Contract() {

  const [hasError, setErrors] = useState(false);

  const [contract, setContract] = useState({});

  const [selectedID, setselectedID] = useState(useParams().contract_id);
  const [selectedInvoiceID, setSelectedInvoiceID] = useState(null);
  const [finalRelations, setFinalRelations] = useState([[]]);
  const [relations, setRelations] = useState([[]]);
  const [finalRelationsAmount, setFinalRelationsAmount] = useState('...');
  const [contractAmount, setContractAmount] = useState('...');


  const [loading, setLoading] = useState(false);
  const [open, setOpen] = useState(false);

  const classes = useStyles();

  const handleClickOpen = (id) => {
    setSelectedInvoiceID(id);
    setOpen(true);
  };

  const handleClose = () => {
    setOpen(false);
  };

  async function fetchData() {
    setLoading(true);
    const res = await fetch(matcher_api_url + "/contracts/" + selectedID);
    res
      .json()
      .then(res => {
          setContract(res.contract);
          setContractAmount(Math.max.apply(Math, [res.contract.amount_with_dph, res.contract.amount_without_dph]));
      })
      .catch(err => setErrors(err));
    const rels = await fetch(matcher_api_url + "/relations/contract/" + selectedID);
    rels
      .json()
      .then(res => {
          setRelations(res.relations.map((i) => [
              i.ministry_name,
              i.supplier_name,
              i.purpose,
              i.date_issue,
              getNumberInPretyString(i.amount),
              i.score,
              <IconButton color="primary" onClick={() => handleClickOpen(i.invoice_id)}>
                <FindInPage/>
              </IconButton>]));

          setFinalRelations(res.final_relations.map((i) => [
              i.ministry_name,
              i.supplier_name,
              i.purpose,
              i.date_issue,
              getNumberInPretyString(i.amount),
              i.score,
              <IconButton color="primary" onClick={() => handleClickOpen(i.invoice_id)}>
                <FindInPage/>
              </IconButton>]));
          setFinalRelationsAmount(res.final_relations.map((i) => i.amount).reduce((a, b) => a+b, 0));
      })
      .catch(err => setErrors(err));
      setLoading(false);
  }

  useEffect(() => {
    fetchData();

  },[]);

  const getNumberInPretyString = (number) => {
    if(number){
      return parseFloat(String(number)).toLocaleString('cs-CZ', {style: "currency", currency: "CZK"})
    } else {
      return ''
    }
  };

  const getDateInPretyString = (value) => {
    if(value){
      return new Date(value).toLocaleDateString().split("/").join(".")
    } else {
      return ''
    }
  };

  const getAttributeArray = (contract) => {
    return [
      {label:"Ministerstvo",              id:"ministerstvo", value:contract.ministry_name},
      {label:"Dodavatel",                 id:"dodavatel", value:contract.supplier_name},
      {label:"IČO Ministerstva",          id:"ministerstvo_ico", value:contract.ministry_ico},
      {label:"IČO Dodavatele",            id:"dodavatel_ico", value:contract.supplier_ico},
      {label:"Adresa ministerstva",       id:"ministerstvo_adresa", value:contract.ministry_address},
      {label:"Adresa dodavatele",         id:"dodavatel_adresa", value:contract.supplier_address},
      {label:"Datová schránka ministerstva", id:"datova_schranka_ministerstva", value:contract.ministry_data_box},
      {label:"Datová schránka dodavatele", id:"datova_schranka_dodavatele", value:contract.supplier_data_box},
      {label:"Částka s DPH",              id:"castka_s_dph", value:getNumberInPretyString(contract.amount_with_dph), md:3},
      {label:"Částka bez DPH",            id:"castka_bez_dph", value:getNumberInPretyString(contract.amount_without_dph), md:3},
      {label:"Částka v cizí měně",        id:"castka_v_cizi_mene", value:getNumberInPretyString(contract.amount_different_currency), md:3},
      {label:"Měna",                      id:"mena", value:getNumberInPretyString(contract.currency), md:3},
      {label:"Datum uzavření",            id:"datum_uzavreni", value:getDateInPretyString(contract.date_agreed)},
      {label:"Datum vystavení v Registru smluv", id:"datum_prijeti", value:getDateInPretyString(contract.date_published)},
      {label:"Číslo smlouvy",             id:"cislo_smlouvy", value:contract.contract_number},
      {label:"Schválil",                  id:"schvalil", value:contract.approved},
      {label:"Link",                      id:"link", value:<Link href={contract.link} target="_blank">{contract.link}</Link>},
      {label:"Platná",                    id:"platna", value:contract.valid},
      {label:"Číslo smlouvy v Registru smluv", id:"cislo_smlouvy_v_registru_smluv", value:contract.external_id},
      {label:"Číslo verze v Registru smluv", id:"cislo_verze_v_registru_smluv", value:contract.version_id},
      {label:"Předmět smlouvy",           id:"predmet_smlouvy", value:contract.purpose, md:12}
    ]
  };

  return (
    <div>
      <GridContainer>
        <GridItem xs={12} sm={12} md={12}>
          <Card>
            <CardHeader color="primary">
              <h4 className={classes.cardTitleWhite}>Smlouva č.{contract.contract_id}</h4>
              <p className={classes.cardCategoryWhite}>{contract.purpose}</p>
            </CardHeader>
            <CardBody>
              <GridContainer>
              {getAttributeArray(contract).map(attribute => (
                  <InfoColumn label={attribute.label} id={attribute.id} value={attribute.value}
                    xs={attribute.xs ?? 12} sm={attribute.sm ?? 12} md={attribute.md ?? 6} />
              ))}
              </GridContainer>

            </CardBody>
          </Card>
        </GridItem>
      </GridContainer>
      <GridContainer>
        <GridItem xs={12} sm={12} md={4}>
          <Card>
            <CardHeader color="success" stats icon>
              <CardIcon color="success">
                <Icon>description</Icon>
              </CardIcon>
              <p className={classes.cardTitleBlack}>Počet navázaných faktur</p>
              <h3 className={classes.cardTitleBlack}>
                {finalRelations.length}
              </h3>
            </CardHeader>
          </Card>
        </GridItem>
        <GridItem xs={12} sm={12} md={4}>
          <Card>
            <CardHeader color={finalRelationsAmount > contractAmount ? "warning" : "success"} stats icon>
              <CardIcon color={finalRelationsAmount > contractAmount ? "warning" : "success"}>
                <Icon>file_copy</Icon>
              </CardIcon>
              <p className={classes.cardTitleBlack}>Hodnota navázaných faktur</p>
              <h3 className={classes.cardTitleBlack}>
                {getNumberInPretyString(finalRelationsAmount)}
              </h3>
            </CardHeader>
          </Card>
        </GridItem>
        <GridItem xs={12} sm={12} md={4}>
          <Card>
            <CardHeader color="success" stats icon>
              <CardIcon color="success">
                <Icon>list_alt</Icon>
              </CardIcon>
              <p className={classes.cardTitleBlack}>Hodnota smlouvy</p>
              <h3 className={classes.cardTitleBlack}>
                {getNumberInPretyString(contractAmount)}
              </h3>
            </CardHeader>
          </Card>
        </GridItem>
      </GridContainer>
      <GridContainer>
        <GridItem xs={12} sm={12} md={12}>
          <Card>
            <CardHeader color="primary">
              <h4 className={classes.cardTitleWhite}>Faktury
              {loading ? <LinearProgress /> : null}
              </h4>
            </CardHeader>
            <CardBody>
              <Table
                tableHeaderColor="primary"
                tableHead={["Ministerstvo", "Dodavatel", "Předmět", "Datum vystavení", "Částka", "Score", "Akce"]}
                tableData={finalRelations}
              />
            </CardBody>
          </Card>
        </GridItem>
      </GridContainer>
      <GridContainer>
        <GridItem xs={12} sm={12} md={12}>
          <Card>
            <CardHeader color="primary">
              <h4 className={classes.cardTitleWhite}>Další faktury
              {loading ? <LinearProgress /> : null}
              </h4>
            </CardHeader>
            <CardBody>
              <Table
                tableHeaderColor="primary"
                tableHead={["Ministerstvo", "Dodavatel", "Předmět", "Datum vystavení", "Částka", "Score", "Akce"]}
                tableData={relations}
              />
            </CardBody>
            <InvoiceDialog invoiceID={selectedInvoiceID} open={open} onClose={handleClose}/>
          </Card>
        </GridItem>
      </GridContainer>
    </div>
  );
}
